pkg/csr/csr.go

```diff
@@ -0,0 +1,260 @@
+func (c *Approver) validateKubeletServingCSR(csr *x509.CertificateRequest) (bool, error) {
+	if c.CertificateSigningRequest == nil || csr == nil {
+		return false, errors.New("CSR or request should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(c.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR %s contains invalid number of groups: %d", c.Name, len(c.Spec.Groups))
+	}
+	groups := sets.NewString(c.Spec.Groups...)
+	if !groups.HasAll(nodeGroup, systemPrefix) {
+		return false, errors.Errorf("CSR %s does not contain required groups", c.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !c.hasUsages(kubeletServerUsages) {
+		return false, errors.Errorf("CSR %s does not contain required usages", c.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR %s does not contain required subject organization", c.Name)
+	}
+	return true, nil
+}
+
+// isNodeClientCert returns true if the CSR is from a  kube-apiserver-client-kubelet signer
+// reference: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers
+func (c *Approver) isNodeClientCert(x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{nodeGroup}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	// Check usages include: digital signature, key encipherment and client auth
+	if !c.hasUsages(kubeletClientUsages) {
+		return false
+	}
+	return true
+}
+
+// hasUsages verifies if the required usages exist in the CSR spec
+func (c *Approver) hasUsages(usages []certificates.KeyUsage) bool {
+	if len(usages) != len(c.Spec.Usages) {
+		return false
+	}
+
+	usageMap := map[certificates.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range c.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+// ParseCSR extracts the CSR from the API object and decodes it.
+func ParseCSR(csr []byte) (*x509.CertificateRequest, error) {
+	if len(csr) == 0 {
+		return nil, fmt.Errorf("CSR request spec should not be empty")
+	}
+	// extract PEM from request object
+	block, _ := pem.Decode(csr)
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
+		return nil, fmt.Errorf("PEM block type must be CERTIFICATE REQUEST")
+	}
+	return x509.ParseCertificateRequest(block.Bytes)
+}
+
+// areAddressesEqual returns true if the DNS address in CSR subject matches with the instance address
+// If the address passed is an IP address, we do a reverse lookup to find the host address
+func (c *Approver) areAddressesEqual(instanceAdd string, nodeName string) bool {
```

Comment:
Would be nice to unit test this logic.

Reply:
`areAddressesEqual` does not convey the intent of this function. You are trying to see if the `nodeName` maps to `instanceAdd`